### PR TITLE
make nix-shell optional

### PR DESCRIPTION
**Pull request by zimbatm**
Reduce most of the scripts startup times by bringing the nix-shell
invocation to the top-level.

When entering the project, first run `nix-shell`, then do all of the
things as usual.

This allows non-nix users to also use the project, if they install all
the corresponding dependencies.

### CODE REVIEW

**Comment by zimbatm** [comment]
I am not totally happy with this solution either. It makes the execution faster but there is now a disconnect between a dependency and it's script. If all the dependencies are listed in shell.nix it's hard to tell for what they are being use for.

I still think that this is an improvement and have plans for something even better in the future.

File: .travis.yml [hunk 1]
```diff
@@ -3,6 +3,6 @@ os:
 - linux
 language: nix
 script:
-  - ./travis/build
-  - ./travis/deploy
-  - ./travis/deploy-cleanup
+  - nix-shell --run ./travis/build
```
Comment (nmattia) [inline .travis.yml]: can you run `--pure` just to make sure nothing is missing and that we don't pick up stuff from the machine environment?

**Comment by zimbatm** [comment]
> there are some tweaks to the deploy script, is it meant to be in this PR as well?

Those were fulfilled by the swarm/shell.nix. Now that it's gone I had to code an imperative version of it. I actually think that it's better in terms of information locality.

**Comment by zimbatm** [comment]
> there are some tweaks to the deploy script, is it meant to be in this PR as well?

Those were fulfilled by the swarm/shell.nix. Now that it's gone I had to code an imperative version of it. I actually think that it's better in terms of information locality.

File: .travis.yml (hunk 1 above)
Comment (zimbatm) [reply .travis.yml]: Maybe a further improvement could be to create a ./travis/enter script that invokes `nix-shell --pure` but restores the travis environment variables by passing them as an argument.

**Pull request merged**